### PR TITLE
refactor(release-ci): bump `cargo-dist` 0.1.0

### PR DESCRIPTION
**Pull request by smallstepman**
Example run of updated CI: https://github.com/smallstepman/dfx-extensions/actions/runs/5879509170
Example release after CI updates: https://github.com/smallstepman/dfx-extensions/releases/tag/sns-v23132123.231.123

Updates release workflow to use cargo-dist 0.1.0: https://github.com/axodotdev/cargo-dist/releases/tag/v0.1.0. Most notable changes:
- cargo-dist compiles the package using the version of rust specified in rust-toolchain.toml
- cargo-dist packages tarball into `tar.gz` (previously it was `tar.xz` which we had to repackage in CI)
- cargo-dist calculates `sha256` for the archive and uploads it as `{archive_filename_with_file_extension}.sha256`
- cargo-dist changed the default naming convention for the tarball archive (from `{package}-v{version}-{platform}.{ext}` to `{package}-{platform}.{ext}`, read more here `https://github.com/axodotdev/cargo-dist/issues/184`)
- fixed a bug that caused a package to get published in crates.io even when `publish = false` was specified in `Cargo.toml`

Additionally:
- we're no longer using `dtolnay/rust-toolchain@stable`; rust toolchain comes preinstalled on GitHub Runners
- `cargo-binstall` manual installation has been replaced with minimal GHA (see `https://github.com/cargo-bins/cargo-binstall/pull/1269`) 
- a few lints here and there

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -110,32 +111,38 @@ jobs:
...
-          install-dist: curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v0.0.5/cargo-dist-v0.0.5-installer.sh | sh
+        - os: "macos-latest"
+          dist-args: "--artifacts=local --target=aarch64-apple-darwin"
+          install-dist: "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v0.1.0/cargo-dist-installer.sh | sh"
+        - os: "macos-latest"
+          dist-args: "--artifacts=local --target=x86_64-apple-darwin"
+          install-dist: "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v0.1.0/cargo-dist-installer.sh | sh"
+        - os: "ubuntu-latest"
+          dist-args: "--artifacts=local --target=x86_64-unknown-linux-gnu"
+          install-dist: "curl --proto '=https' --tlsv1.2 -LsSf https://github.com/axodotdev/cargo-dist/releases/download/v0.1.0/cargo-dist-installer.sh | sh"
```
Comment (ericswanson-dfinity) [inline .github/workflows/release.yml:126]: What's the purpose of defining `install-dist` in the matrix?  Unless I'm missing something, it is identical for all variants.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -1,3 +1,6 @@
+# Copyright 2022-2023, axodotdev
+# SPDX-License-Identifier: MIT or Apache-2.0
```
Comment (ericswanson-dfinity) [inline .github/workflows/release.yml:2]: This copyright applies to what?  The whole workflow?

File: .github/workflows/release-with-github.yml [hunk 3]
```diff
@@ -43,16 +43,8 @@ jobs:
...
-      - name: Install dependencies
-        run: |
-          wget https://github.com/cargo-bins/cargo-binstall/releases/latest/download/cargo-binstall-x86_64-unknown-linux-musl.tgz
-          tar -xvzf cargo-binstall-x86_64-unknown-linux-musl.tgz
-          rm cargo-binstall-x86_64-unknown-linux-musl.tgz
-          mv cargo-binstall $HOME/.cargo/bin
-          cargo binstall cargo-release -y
-          cargo binstall ripgrep -y
+      - uses: cargo-bins/cargo-binstall@main
+      - run: cargo binstall cargo-release ripgrep -y
```
Comment (ericswanson-dfinity) [inline .github/workflows/release-with-github.yml:47]: Does this workflow use some special features of ripgrep that grep doesn't have?  If not, we should use grep.

Comment (smallstepman) [reply .github/workflows/release-with-github.yml:47]: Capture groups, which AFAIK, work differently on MacOS and Linux. Installing both packages took 1.983274393s. I'd appreciate it if we could leave it as is for the sake of debuggability

File: .github/workflows/release.yml (hunk 2 above)
Comment (smallstepman) [reply .github/workflows/release.yml:2]: It came out from `cargo dist generate-ci` like that. I think we can remove it since it's MIT?

Comment (smallstepman) [reply .github/workflows/release.yml:2]: no, I don't think we should remove it, but I don't see a problem with it either:
- https://github.com/axodotdev/cargo-dist/blob/main/LICENSE-MIT
- https://github.com/axodotdev/cargo-dist/blob/main/LICENSE-APACHE



## what LLM has to say
Removing a copyright notice and license identifier from a source code file may have legal consequences, depending on various factors, including the license terms under which the software was originally distributed. Let’s break down the parts of this notice:

Copyright Notice: This asserts ownership of the intellectual property. It is a legal statement by the author/owner that they hold the copyright to the material in the file.

SPDX-License-Identifier: This is a standard way to express licensing information. In this example, it is indicating that the code is licensed under either the MIT License or the Apache 2.0 License. Both of these are permissive open-source licenses, but they have terms that must be followed.

Here are some things to consider:

Legally Required: For many open source licenses, including MIT and Apache 2.0, you are generally required to keep the copyright notice and license text with the software, even as you modify it. Removing these notices might put you in violation of the license terms.

Ethical Considerations: Even if not strictly legally required (which is rare with open-source licenses), keeping the copyright notice is often considered the ethical thing to do. It gives credit to the original author(s) of the code.

Potential Consequences: If you remove the notice and distribute the software (either in source or binary form), you may be doing so without a license, which means you could be infringing on the original author's copyright. This could potentially lead to legal action against you.

Modification for Personal Use: If you are modifying the source code for your personal use and have no intention of distributing it, the situation might be different. Some licenses are more lenient in this respect, but it’s still a best practice to retain the notices.

Check the License Terms: The terms of the MIT and Apache 2.0 licenses generally require you to include the copyright notice and license text when you redistribute the code, but you should read these licenses in detail to understand their terms. Other licenses may have different requirements.

Getting Permission: If you have a legitimate reason for wanting to remove or modify the notice, consider reaching out to the copyright holder (in this case, presumably "axodotdev") and asking for permission.

In summary, before you remove such notices, you should consult the terms of the licenses involved and, if necessary, consult with a legal professional. It is usually best practice—and often a legal requirement—to keep these notices intact.
